File: rust/src/daemon_client.rs
//! Daemon client utilities for index access.

use std::collections::BTreeMap;
use std::env;
use std::io::{BufRead, BufReader, Write};
#[cfg(unix)]
use std::os::unix::net::UnixStream;
use std::path::Path;
use std::process::{Command, Stdio};
use std::sync::{Mutex, OnceLock};
use std::time::Duration;

use serde_json::Value;
use uuid::Uuid;

use crate::daemon_paths::get_daemon_socket_path;
use crate::daemon_protocol::{ErrorEnvelope, RequestEnvelope, ResponseEnvelope, PROTOCOL_VERSION};
use crate::error::TaskulusError;

/// Test-only response override for daemon client requests.
#[derive(Clone, Debug)]
pub enum TestDaemonResponse {
    /// Simulate an empty daemon response.
    Empty,
    /// Simulate a daemon connection error.
    IoError,
    /// Return a fixed response envelope.
    Envelope(ResponseEnvelope),
}

static TEST_DAEMON_RESPONSES: OnceLock<Mutex<Vec<TestDaemonResponse>>> = OnceLock::new();
static TEST_DAEMON_SPAWN_DISABLED: OnceLock<Mutex<bool>> = OnceLock::new();

/// Set the test response for the next daemon request.
///
/// Passing `None` clears any pending override.
pub fn set_test_daemon_response(response: Option<TestDaemonResponse>) {
    let cell = TEST_DAEMON_RESPONSES.get_or_init(|| Mutex::new(Vec::new()));
    let mut guard = cell.lock().expect("lock test response");
    guard.clear();
    if let Some(item) = response {
        guard.push(item);
    }
}

/// Set a sequence of test responses for daemon requests.
pub fn set_test_daemon_responses(responses: Vec<TestDaemonResponse>) {
    let cell = TEST_DAEMON_RESPONSES.get_or_init(|| Mutex::new(Vec::new()));
    let mut guard = cell.lock().expect("lock test responses");
    *guard = responses;
}

/// Return whether a test daemon response override is set.
pub fn has_test_daemon_response() -> bool {
    let cell = TEST_DAEMON_RESPONSES.get_or_init(|| Mutex::new(Vec::new()));
    let guard = cell.lock().expect("lock test response");
    !guard.is_empty()
}

/// Disable daemon spawning for tests when set to true.
pub fn set_test_daemon_spawn_disabled(disabled: bool) {
    let cell = TEST_DAEMON_SPAWN_DISABLED.get_or_init(|| Mutex::new(false));
    let mut guard = cell.lock().expect("lock test spawn flag");
    *guard = disabled;
}

fn take_test_daemon_response() -> Option<TestDaemonResponse> {
    let cell = TEST_DAEMON_RESPONSES.get_or_init(|| Mutex::new(Vec::new()));
    let mut guard = cell.lock().expect("lock test response");
    if guard.is_empty() {
        None
    } else {
        Some(guard.remove(0))
    }
}

fn is_test_spawn_disabled() -> bool {
    let cell = TEST_DAEMON_SPAWN_DISABLED.get_or_init(|| Mutex::new(false));
    let guard = cell.lock().expect("lock test spawn flag");
    *guard
}

/// Return whether daemon mode is enabled.
pub fn is_daemon_enabled() -> bool {
    let value = env::var("TASKULUS_NO_DAEMON")
        .unwrap_or_default()
        .to_lowercase();
    !matches!(value.as_str(), "1" | "true" | "yes")
}

/// Request index list from the daemon, spawning it if needed.
///
/// # Arguments
/// * `root` - Repository root path.
///
/// # Errors
/// Returns `TaskulusError` if daemon request fails.
pub fn request_index_list(root: &Path) -> Result<Vec<Value>, TaskulusError> {
    if !is_daemon_enabled() {
        return Err(TaskulusError::IssueOperation("daemon disabled".to_string()));
    }
    let socket_path = get_daemon_socket_path(root)?;
    let request = RequestEnvelope {
        protocol_version: PROTOCOL_VERSION.to_string(),
        request_id: format!("req-{}", Uuid::new_v4().simple()),
        action: "index.list".to_string(),
        payload: BTreeMap::new(),
    };
    if !socket_path.exists() {
        spawn_daemon(root)?;
    }
    let response = request_with_recovery(&socket_path, &request, root)?;
    if response.status != "ok" {
        let error = response.error.unwrap_or(ErrorEnvelope {
            code: "internal_error".to_string(),
            message: "daemon error".to_string(),
            details: BTreeMap::new(),
        });
        return Err(TaskulusError::IssueOperation(error.message));
    }
    let result = response.result.unwrap_or_default();
    match result.get("issues") {
        Some(Value::Array(values)) => Ok(values.clone()),
        _ => Ok(Vec::new()),
    }
}

/// Request daemon status.
pub fn request_status(root: &Path) -> Result<BTreeMap<String, Value>, TaskulusError> {
    if !is_daemon_enabled() {
        return Err(TaskulusError::IssueOperation("daemon disabled".to_string()));
    }
    let socket_path = get_daemon_socket_path(root)?;
    let request = RequestEnvelope {
        protocol_version: PROTOCOL_VERSION.to_string(),
        request_id: format!("req-{}", Uuid::new_v4().simple()),
        action: "ping".to_string(),
        payload: BTreeMap::new(),
    };
    let response = request_with_recovery(&socket_path, &request, root)?;
    if response.status != "ok" {
        let error = response.error.unwrap_or(ErrorEnvelope {
            code: "internal_error".to_string(),
            message: "daemon error".to_string(),
            details: BTreeMap::new(),
        });
        return Err(TaskulusError::IssueOperation(error.message));
    }
    Ok(response.result.unwrap_or_default())
}

/// Request daemon shutdown.
pub fn request_shutdown(root: &Path) -> Result<BTreeMap<String, Value>, TaskulusError> {
    if !is_daemon_enabled() {
        return Err(TaskulusError::IssueOperation("daemon disabled".to_string()));
    }
    let socket_path = get_daemon_socket_path(root)?;
    let request = RequestEnvelope {
        protocol_version: PROTOCOL_VERSION.to_string(),
        request_id: format!("req-{}", Uuid::new_v4().simple()),
        action: "shutdown".to_string(),
        payload: BTreeMap::new(),
    };
    let response = request_with_recovery(&socket_path, &request, root)?;
    if response.status != "ok" {
        let error = response.error.unwrap_or(ErrorEnvelope {
            code: "internal_error".to_string(),
            message: "daemon error".to_string(),
            details: BTreeMap::new(),
        });
        return Err(TaskulusError::IssueOperation(error.message));
    }
    Ok(response.result.unwrap_or_default())
}

fn request_with_recovery(
    socket_path: &Path,
    request: &RequestEnvelope,
    root: &Path,
) -> Result<ResponseEnvelope, TaskulusError> {
    match send_request(socket_path, request) {
        Ok(response) => Ok(response),
        Err(error) => {
            if !matches!(error, TaskulusError::Io(_)) {
                return Err(error);
            }
            if socket_path.exists() {
                std::fs::remove_file(socket_path)
                    .map_err(|error| TaskulusError::Io(error.to_string()))?;
            }
            spawn_daemon(root)?;
            let mut last_error = error;
            for _ in 0..10 {
                match send_request(socket_path, request) {
                    Ok(response) => return Ok(response),
                    Err(err) => {
                        if !matches!(err, TaskulusError::Io(_)) {
                            return Err(err);
                        }
                        last_error = err;
                        std::thread::sleep(Duration::from_millis(50));
                    }
                }
            }
            Err(last_error)
        }
    }
}

#[cfg(unix)]
fn send_request(
    socket_path: &Path,
    request: &RequestEnvelope,
) -> Result<ResponseEnvelope, TaskulusError> {
    if let Some(response) = take_test_daemon_response() {
        return match response {
            TestDaemonResponse::Empty => Err(TaskulusError::IssueOperation(
                "empty daemon response".to_string(),
            )),
            TestDaemonResponse::IoError => {
                Err(TaskulusError::Io("daemon connection failed".to_string()))
            }
            TestDaemonResponse::Envelope(envelope) => Ok(envelope),
        };
    }
    let mut stream =
        UnixStream::connect(socket_path).map_err(|error| TaskulusError::Io(error.to_string()))?;
    let payload =
        serde_json::to_string(request).map_err(|error| TaskulusError::Io(error.to_string()))?;
    stream
        .write_all(payload.as_bytes())
        .map_err(|error| TaskulusError::Io(error.to_string()))?;
    stream
        .write_all(b"\n")
        .map_err(|error| TaskulusError::Io(error.to_string()))?;
    let mut reader = BufReader::new(stream);
    let mut line = String::new();
    reader
        .read_line(&mut line)
        .map_err(|error| TaskulusError::Io(error.to_string()))?;
    if line.trim().is_empty() {
        return Err(TaskulusError::IssueOperation(
            "empty daemon response".to_string(),
        ));
    }
    serde_json::from_str(&line).map_err(|error| TaskulusError::Io(error.to_string()))
}

#[cfg(not(unix))]
fn send_request(
    _socket_path: &Path,
    _request: &RequestEnvelope,
) -> Result<ResponseEnvelope, TaskulusError> {
    Err(TaskulusError::IssueOperation(
        "daemon not supported on this platform".to_string(),
    ))
}

#[cfg(unix)]
fn spawn_daemon(root: &Path) -> Result<(), TaskulusError> {
    if is_test_spawn_disabled() {
        return Ok(());
    }
    Command::new(std::env::current_exe().map_err(|error| TaskulusError::Io(error.to_string()))?)
        .arg("daemon")
        .arg("--root")
        .arg(root)
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn()
        .map_err(|error| TaskulusError::Io(error.to_string()))?;
    Ok(())
}

#[cfg(not(unix))]
fn spawn_daemon(_root: &Path) -> Result<(), TaskulusError> {
    Err(TaskulusError::IssueOperation(
        "daemon not supported on this platform".to_string(),
    ))
}

File: python/src/taskulus/daemon_client.py
"""Daemon client utilities for index access."""

from __future__ import annotations

import json
import os
import socket
import subprocess
import sys
import uuid
import time
from pathlib import Path
from typing import Any, Dict, List

from taskulus.daemon_paths import get_daemon_socket_path
from taskulus.daemon_protocol import (
    PROTOCOL_VERSION,
    ErrorEnvelope,
    RequestEnvelope,
    ResponseEnvelope,
)


class DaemonClientError(RuntimeError):
    """Raised when daemon communication fails."""


def is_daemon_enabled() -> bool:
    """Return whether daemon mode is enabled.

    :return: True when daemon mode is enabled.
    :rtype: bool
    """
    value = os.getenv("TASKULUS_NO_DAEMON", "").lower()
    return value not in {"1", "true", "yes"}


def send_request(socket_path: Path, request: RequestEnvelope) -> ResponseEnvelope:
    """Send a request to the daemon and return the response.

    :param socket_path: Daemon socket path.
    :type socket_path: Path
    :param request: Request envelope.
    :type request: RequestEnvelope
    :return: Response envelope.
    :rtype: ResponseEnvelope
    :raises DaemonClientError: If communication fails.
    """
    payload = json.dumps(request.model_dump(mode="json")).encode("utf-8") + b"\n"
    try:
        with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
            sock.settimeout(2.0)
            sock.connect(str(socket_path))
            sock.sendall(payload)
            response_raw = sock.makefile("rb").readline()
    except OSError as error:
        raise DaemonClientError("daemon connection failed") from error

    if not response_raw:
        raise DaemonClientError("empty daemon response")
    response_payload = json.loads(response_raw.decode("utf-8"))
    return ResponseEnvelope.model_validate(response_payload)


def spawn_daemon(root: Path) -> None:
    """Spawn the daemon process.

    :param root: Repository root path.
    :type root: Path
    """
    subprocess.Popen(
        [sys.executable, "-m", "taskulus.daemon", "--root", str(root)],
        cwd=root,
        start_new_session=True,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )


def request_index_list(root: Path) -> List[Dict[str, Any]]:
    """Request the index list from the daemon, spawning it if needed.

    :param root: Repository root path.
    :type root: Path
    :return: List of issue payloads.
    :rtype: List[Dict[str, Any]]
    :raises DaemonClientError: If the daemon request fails.
    """
    if not is_daemon_enabled():
        raise DaemonClientError("daemon disabled")
    socket_path = get_daemon_socket_path(root)
    request = RequestEnvelope(
        protocol_version=PROTOCOL_VERSION,
        request_id=f"req-{uuid.uuid4().hex[:8]}",
        action="index.list",
        payload={},
    )
    if not socket_path.exists():
        spawn_daemon(root)
    response = _request_with_recovery(socket_path, request, root)
    if response.status != "ok":
        error = response.error or ErrorEnvelope(
            code="internal_error", message="daemon error", details={}
        )
        raise DaemonClientError(error.message)
    result = response.result or {}
    return list(result.get("issues", []))


def request_status(root: Path) -> Dict[str, Any]:
    """Request daemon status.

    :param root: Repository root path.
    :type root: Path
    :return: Status payload.
    :rtype: Dict[str, Any]
    """
    if not is_daemon_enabled():
        raise DaemonClientError("daemon disabled")
    socket_path = get_daemon_socket_path(root)
    request = RequestEnvelope(
        protocol_version=PROTOCOL_VERSION,
        request_id=f"req-{uuid.uuid4().hex[:8]}",
        action="ping",
        payload={},
    )
    response = _request_with_recovery(socket_path, request, root)
    if response.status != "ok":
        error = response.error or ErrorEnvelope(
            code="internal_error", message="daemon error", details={}
        )
        raise DaemonClientError(error.message)
    return response.result or {}


def request_shutdown(root: Path) -> Dict[str, Any]:
    """Request daemon shutdown.

    :param root: Repository root path.
    :type root: Path
    :return: Shutdown response payload.
    :rtype: Dict[str, Any]
    """
    if not is_daemon_enabled():
        raise DaemonClientError("daemon disabled")
    socket_path = get_daemon_socket_path(root)
    request = RequestEnvelope(
        protocol_version=PROTOCOL_VERSION,
        request_id=f"req-{uuid.uuid4().hex[:8]}",
        action="shutdown",
        payload={},
    )
    response = _request_with_recovery(socket_path, request, root)
    if response.status != "ok":
        error = response.error or ErrorEnvelope(
            code="internal_error", message="daemon error", details={}
        )
        raise DaemonClientError(error.message)
    return response.result or {}


def _request_with_recovery(
    socket_path: Path, request: RequestEnvelope, root: Path
) -> ResponseEnvelope:
    try:
        return send_request(socket_path, request)
    except DaemonClientError as error:
        if str(error) != "daemon connection failed":
            raise
        if socket_path.exists():
            socket_path.unlink()
        spawn_daemon(root)
        last_error = error
        for _ in range(10):
            try:
                return send_request(socket_path, request)
            except DaemonClientError as retry_error:
                if str(retry_error) != "daemon connection failed":
                    raise
                last_error = retry_error
                time.sleep(0.05)
        raise last_error
